plugins/thunderbolt/fu-plugin-thunderbolt.c

```diff
@@ -0,0 +1,736 @@
+				   guint     timeout_ms)
+{
+	FuPluginData *data = fu_plugin_get_data (plugin);
+	data->timeout = timeout_ms;
+}
+
+/* virtual functions */
+
+void
+fu_plugin_init (FuPlugin *plugin)
+{
+	FuPluginData *data = fu_plugin_alloc_data (plugin, sizeof (FuPluginData));
+	const gchar *subsystems[] = { "thunderbolt", NULL };
+
+	data->udev = g_udev_client_new (subsystems);
+	g_signal_connect (data->udev, "uevent",
+			  G_CALLBACK (udev_uevent_cb), plugin);
+
+	data->timeout = FU_PLUGIN_THUNDERBOLT_UPDATE_TIMEOUT_MS;
+}
+
+void
+fu_plugin_destroy (FuPlugin *plugin)
+{
+	FuPluginData *data = fu_plugin_get_data (plugin);
+	g_object_unref (data->udev);
+}
+
+gboolean
+fu_plugin_coldplug (FuPlugin *plugin, GError **error)
+{
+	FuPluginData *data = fu_plugin_get_data (plugin);
+	GList *devices, *l;
+
+	devices = g_udev_client_query_by_subsystem (data->udev, "thunderbolt");
+	for (l = devices; l != NULL; l = l->next) {
+		GUdevDevice *device = l->data;
+		fu_plugin_thunderbolt_add (plugin, device);
+	}
+
+	g_list_foreach (devices, (GFunc) g_object_unref, NULL);
+	g_list_free (devices);
+
+	return TRUE;
+}
+
+
+gboolean
+fu_plugin_update_online (FuPlugin *plugin,
+			 FuDevice *dev,
+			 GBytes *blob_fw,
+			 FwupdInstallFlags flags,
+			 GError **error)
+{
+	FuPluginData *data = fu_plugin_get_data (plugin);
+	const gchar *devpath;
+	gboolean ret;
+	guint64 status;
+	g_autoptr(GUdevDevice) udevice = NULL;
+	g_autoptr(GError) error_local = NULL;
+
+	ret = fu_plugin_thunderbolt_validate_firmware (blob_fw, &error_local);
+	if (!ret) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_FOUND,
+			     "could not validate firmware: %s",
+			     error_local->message);
+		return FALSE;
+	}
+
+	devpath = fu_device_get_metadata (dev, "sysfs-path");
+	g_return_val_if_fail (devpath, FALSE);
+
+	udevice = g_udev_client_query_by_sysfs_path (data->udev, devpath);
+	if (udevice == NULL) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_FOUND,
+			     "could not find thunderbolt device at %s",
+			     devpath);
+		return FALSE;
+	}
+
+	fu_plugin_set_status (plugin, FWUPD_STATUS_DEVICE_WRITE);
+	ret = fu_plugin_thunderbolt_write_firmware (udevice, blob_fw, &error_local);
+	if (!ret) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_WRITE,
+			     "could not write firmware to thunderbolt device at %s: %s",
+			     devpath, error_local->message);
+		return FALSE;
+	}
+
+	ret = fu_plugin_thunderbolt_trigger_update (udevice, &error_local);
+	if (!ret) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "Could not start thunderbolt device upgrade: %s",
+			     error_local->message);
+		return FALSE;
+	}
+
+	fu_plugin_set_status (plugin, FWUPD_STATUS_DEVICE_RESTART);
+
+	/* the device will disappear and we need to wait until it reappears,
+	 * and then check if we find an error */
+	ret = fu_plugin_thunderbolt_wait_for_device (plugin,
```

Comment:
@ybernat so I just checked again (upading the host on the XPS 13, i.e. `domain0/0-0`. This is my log: https://paste.fedoraproject.org/paste/nH4xa3lxmKI6SRblpAn~vA
I see the domain being removed, added, removed and added. But not the device that was updated (`domain/0-0`). Is that what you would expect?